pkg/api-manager/istiohandler.go

```diff
@@ -0,0 +1,264 @@
+	// InternalGateway represents the gateway that we should redirect to
+	// Should probably come from a config file
+	InternalGateway = "knative-ingressgateway.istio-system.svc.cluster.local"
+)
+
+func makeMethodRegex(methods []string) string {
+	return strings.Join(methods, "|") + "/g"
+}
+
+func makeMethodsFromRegex(regex string) []string {
+	split := strings.Split(regex, "|")
+	if len(split) == 1 {
+		return []string{strings.Split(split[0], "/g")[0]}
+	}
+	return split[0 : len(split)-1]
+}
+
+func apiModelOntoIstioEntity(m *v1.API) VirtualService {
+	cors := CorsPolicy{}
+	if m.Cors {
+		cors.AllowMethods = m.Methods
+	}
+	var routes []HttpRoute
+	for _, prefix := range m.Uris {
+		if prefix == "/" {
+			prefix = fmt.Sprintf("/%v", *m.Function)
+		}
+		match := HttpMatch{
+			URI:    URI{Prefix: prefix},
+			Method: MethodMatch{Regex: makeMethodRegex(m.Methods)},
+		}
+		route := HttpRoute{
+			Match: []HttpMatch{match},
+			Route: []DestWeight{DestWeight{
+				Destination: RouteDest{
+					Host: InternalGateway,
+				},
+				Weight: 100,
+			}},
+			Rewrite: HttpRewrite{
+				Authority: fmt.Sprintf("%v.default.example.com", *m.Function),
+			},
+			CORS: cors,
+		}
+		routes = append(routes, route)
+	}
+
+	hosts := m.Hosts
+	if len(m.Hosts) == 0 {
+		hosts = append(hosts, "*")
+	}
+
+	return VirtualService{
+		Name:     *m.Name,
+		Function: *m.Function,
+		Hosts:    hosts,
+		Gateways: []string{"knative-shared-gateway.knative-serving.svc.cluster.local"},
+		HTTP:     routes,
+	}
+}
+
+func corsEnabled(cors CorsPolicy) bool {
+	if len(cors.AllowOrigin) != 0 {
+		return true
+	} else if len(cors.AllowMethods) != 0 {
+		return true
+	} else if len(cors.AllowHeaders) != 0 {
+		return true
+	}
+	return false
+}
+
+func istioEntityToModel(vs VirtualService) *v1.API {
+	var uris []string
+	var methods []string
+	var function string
+	var cors bool
+	hosts := vs.Hosts
+	for _, route := range vs.HTTP {
+		log.Infof("Route: %+v", route)
+		uris = append(uris, route.Match[0].URI.Prefix)
+		methods = makeMethodsFromRegex(route.Match[0].Method.Regex)
+		function = strings.Split(route.Rewrite.Authority, ".")[0]
+		if corsEnabled(route.CORS) {
+			cors = true
+		}
+	}
+	log.Infof("Virtual Service: %+v", vs)
+	log.Infof("Methods: %v", methods)
+	return &v1.API{
+		Enabled:   true,
+		Methods:   methods,
+		Name:      &vs.Name,
+		Uris:      uris,
+		Hosts:     hosts,
+		Function:  &function,
+		Protocols: []string{"http"},
+		Cors:      cors,
+	}
+}
+
+type IstioHandlers struct {
+	client *istio.Client
+}
+
+func NewIstioHandlers(client *istio.Client) *IstioHandlers {
+	return &IstioHandlers{
+		client: client,
+	}
+}
+
+func (cl *IstioHandlers) AddAPI(params endpoint.AddAPIParams, principal interface{}) middleware.Responder {
+	span, ctx := trace.Trace(params.HTTPRequest.Context(), "")
+	defer span.Finish()
+
+	log.Infof("Trying to add istio api: %+v", params)
+
+	api := apiModelOntoIstioEntity(params.Body)
+	virtualServiceSpec, err := yaml.Marshal(api)
```

Comment:
Separate Istio data model and YAML marshaling/unmarshaling seems redundant: why not just use istioClient directly?